### ISSUE TITLE
Failure when updating Lenovo Thinkpad Thunderbolt 3 Dock

### ISSUE THREAD

**Issue by thescouser89** [fwupd/fwupd#3171]
**Describe the bug**
While trying to update the Lenovo Thinkpad Thunderbolt 3 Dock (Gen 2) via `fwupdmgr` it failed with:

```
~ ❯ sudo fwupdmgr update         
Devices with no available firmware updates: 
 • UEFI Device Firmware
 • UEFI Device Firmware
 • UEFI Device Firmware
 • UEFI dbx
 • USB2.0 Hub
Devices with the latest available firmware version:
 • Thunderbolt host controller
Upgrade available for USB3.1 Hub from 51.13 to 51.63
USB3.1 Hub and all connected devices may not be usable while updating. Continue with update? [Y|n]: Y
Downloading…             [***************************************]
Decompressing…           [***************************************]
Authenticating…          [***************************************]
Authenticating…          [***************************************]
Updating USB3.1 Hub…     [***************************************]
Erasing…                 [***************************************]
failed to erase FW sector @0x11000: failed to read back empty: failed to read SPI data @0x11be0: transfer failed
```

Original comment: https://github.com/fwupd/firmware-lenovo-thinkpad/issues/5#issuecomment-823668208

@hughsie replied from my comment with this:
```
This is the VLI USB hub and not the MCU -- and shouldn't have damaged anything on failure.
```
https://github.com/fwupd/firmware-lenovo-thinkpad/issues/5#issuecomment-823926074

**Steps to Reproduce**
Run `sudo fwupdmgr update`

**Expected behavior**
The update succeeds

**fwupd version information**
Please provide the version of the daemon and client.

(Edit: updated from original description with latest `dnf update`. Error is the same though)
```shell
~ ❯ fwupdmgr --version
client version:	1.5.9
compile-time dependency versions
	gusb:	0.3.6

daemon version:	1.5.9
~ ❯ 

```

Please note how you installed it (`apt`, `dnf`, `pacman`, source, etc): dnf

**fwupd device information**
Please provide the output of the fwupd devices recognized in your system.

(Edit: I updated the output below with my docking station plugged in)
```shell
~ ❯ fwupdmgr get-devices --show-all-devices                                                                                                                                            
20NYS7K90Z
│
├─Thunderbolt host controller:
│     Device ID:          ccf24df0101fff1f2117a00dbbd2e82d98060109
│     Summary:            Unmatched performance for high-speed I/O
│     Current version:    20.00
│     Vendor:             Lenovo (THUNDERBOLT:0x0109|TBT:0x0109)
│     GUIDs:              ea3c7ac8-d937-568a-aab5-e053bd6af9ce
│                         3d3f953a-3408-56fa-81f3-042d73c59c16
│                         e773c51e-a20c-5b29-9f09-6bb0e0ef7560
│                         f1793149-6c12-5a7f-95db-9fefd116748d
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Supported on remote server
│                         • Device stages updates
│   
├─USB3.1 Hub:
│     Device ID:          54f0d9041b6c5438c7ff825f5139559c5ca1b222
│     Summary:            USB 3.x Hub
│     Current version:    51.13
│     Vendor:             VIA Labs, Inc. (USB:0x17EF)
│     Install Duration:   15 seconds
│     GUIDs:              8ee94f0e-9b44-596a-bdd9-6f90401664cc
│                         4ad676ff-8529-5cb0-930d-e3a1da2c60cb
│                         8aab7e9b-2ae8-559b-9be0-49bf5a647a2a
│                         a9866465-e1ec-5ce0-855d-162d817aeca2
│                         06fb2b19-690e-5296-8d0a-6e32a6fccd84
│     Device Flags:       • Updatable
│                         • Supported on remote server
│                         • Cryptographic hash verification is available
│                         • Device stages updates
│                         • Device can recover flash failures
│   
├─Embedded Controller:
│     Device ID:          09ce7908b2d90a3b53473c3f328410590a527f29
│     Current version:    0.1.17
│     Minimum Version:    0.1.17
│     Vendor:             DMI:LENOVO
│     GUIDs:              be9a5f93-930b-40bf-8864-815268ea4ea3
│                         98129d04-2653-5233-89f1-8aa461e54a25
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Supported on remote server
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│   
├─Intel Management Engine:
│     Device ID:          943ac8e39266b524856ff8e415f05b6f2bcbb148
│     Current version:    192.72.1757
│     Minimum Version:    0.0.1
│     Vendor:             DMI:LENOVO
│     GUIDs:              aeac19da-0c78-4b86-a718-1a043947d83b
│                         78502a54-be51-5aef-82ad-2d640cc19f9b
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Supported on remote server
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│   
├─Intel(R) Core™ i7-8665U CPU @ 1.90GHz:
│     Device ID:          4bde70ba4e39b28f9eab1628f9dd6e6244c03027
│     Current version:    0x000000de
│     Vendor:             Intel
│     GUIDs:              b9a2dd81-159e-5537-a7db-e7101d164d3f
│                         30249f37-d140-5d3e-9319-186b1bd5cac3
│                         561403e8-143a-5071-ab09-bf5e4c146983
│                         d5092916-1868-5182-8c70-12f83dab6372
│     Device Flags:       • Internal device
│   
├─KXG6AZNV256G TOSHIBA:
│     Device ID:          f2759da7fe8e0388c5f3601cb072f837b1070b03
│     Summary:            NVM Express Solid State Drive
│     Current version:    5108AGLA
│     Vendor:             Toshiba Corporation (NVME:0x1179)
│     GUIDs:              9da14130-275e-52fc-8119-63be32bd58dd
│                         141102bc-3608-5bc8-a700-983f47d476f4
│                         de51500a-8b48-5f13-b0c1-0bbded7c54c1
│                         94eae9cd-5ee6-5b80-9a3e-ffa2db88cd87
│                         e04d75fa-e2db-5d35-8549-f971acda90e8
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Supported on remote server
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│   
├─System Firmware:
│ │   Device ID:          ca8e899e46f1da1776d16410bc92f2cf0d963433
│ │   Current version:    0.1.71
│ │   Vendor:             LENOVO (DMI:LENOVO)
│ │   GUIDs:              6300ba52-d9ea-4ca0-88f9-f89ef73cc07b
│ │                       230c8b18-8d9b-53ec-838b-6cfc0383493a
│ │                       df8fecf2-0e94-5191-9b0b-cc3aee5da5c0
│ │   Device Flags:       • Internal device
│ │                       • Updatable
│ │                       • System requires external power source
│ │                       • Supported on remote server
│ │                       • Needs a reboot after installation
│ │                       • Cryptographic hash verification is available
│ │                       • Device is usable for the duration of the update
│ │ 
│ ├─Intel AMT [unprovisioned]:
│ │     Device ID:        e2623122c99d58220498aacbfcfdb1baebbae3c5
│ │     Summary:          Hardware and firmware technology for remote out-of-band management
│ │     Current version:  12.0.72.1757
│ │     Bootloader Version:12.0.72.1757
│ │     Vendor:           Intel Corporation
│ │     GUID:             2800f812-b7b4-2d4b-aca8-46e0ff65814c
│ │     Device Flags:     • Internal device
│ │   
│ └─UEFI dbx:
│       Device ID:        362301da643102b9f38477387e2193e57abaa590
│       Summary:          UEFI Revocation Database
│       Current version:  77
│       Minimum Version:  77
│       Vendor:           UEFI:Linux Foundation
│       Install Duration: 1 second
│       GUIDs:            14503b3d-73ce-5d06-8137-77c68972a341
│                         5971a208-da00-5fce-b5f5-1234342f9cf7
│                         c6682ade-b5ec-57c4-b687-676351208742
│                         f8ba2887-9411-5c36-9cee-88995bb39731
│       Device Flags:     • Internal device
│                         • Updatable
│                         • Needs a reboot after installation
│     
├─TPM:
│ │   Device ID:          c6a80ac3a22083423992a3cb15018989f37834d6
│ │   Current version:    73.8.17568.5511
│ │   Vendor:             ST Microelectronics (TPM:STM)
│ │   GUIDs:              ff71992e-52f7-5eea-94ef-883e56e034c6
│ │                       84df3581-f896-54d2-bd1a-372602f04c32
│ │                       bfaed10a-bbc1-525b-a329-35da2f63e918
│ │                       70b7b833-7e1a-550a-a291-b94a12d0f319
│ │                       06f005e9-cb62-5d1a-82d9-13c534c53c48
│ │   Device Flags:       • Internal device
│ │ 
│ └─Event Log:
│       Device ID:        58bd405f31c48e6eca290b425f530a94c91e955c
│       GUID:             a25657fe-b5dc-5be0-8b78-8b9dfec678ff
│       Device Flags:     • Internal device
│     
├─ThinkPad Thunderbolt 3 Dock USB Audio:
│     Device ID:          962f781b0135936db11c15d09c22c5a4012d4df5
│     Summary:            CX20985 USB audio device
│     Current version:    04-0E-90
│     Bootloader Version: 02.00.00.00
│     Vendor:             Lenovo (USB:0x17EF)
│     Install Duration:   3 seconds
│     GUIDs:              b106286b-108d-5708-a4bd-dff70ce90eea
│                         1d987aa4-e89a-5bd3-b78c-5ec8bdc6e213
│                         d66bf84b-c3ba-508c-bc55-0d445413d3d4
│                         7083139e-9fec-5fc3-9e1a-a0fb3ae1bccf
│     Device Flags:       • Updatable
│                         • Supported on remote server
│   
├─UEFI Device Firmware:
│     Device ID:          0a6ca0e0099a5836e3dec72fd5c67594f3fd9e2b
│     Current version:    16851095
│     Minimum Version:    1
│     Vendor:             DMI:LENOVO
│     GUIDs:              89aab014-a71d-4cae-8824-174808266539
│                         93a0d53f-36ad-5377-a8db-4a618cf4f660
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│   
├─UEFI Device Firmware:
│     Device ID:          b028a468cd8bdc225e88bdb99988cd7646b2687f
│     Current version:    4784136
│     Minimum Version:    4784136
│     Vendor:             DMI:LENOVO
│     GUIDs:              ee246b7d-a882-4e0e-984f-23ac1b0270db
│                         b700928d-20b0-5155-b89e-589b1b70e75f
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│   
├─UEFI Device Firmware:
│     Device ID:          12c8c08f3386ac483cfb6b2908244b82d061b3fe
│     Current version:    26406
│     Minimum Version:    1
│     Vendor:             DMI:LENOVO
│     GUIDs:              9bc92a43-6523-4213-9ccf-4181c849d33d
│                         27a6334c-c2e6-5722-aa85-8d6a90abd6bf
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│   
├─UHD Graphics 620 (Whiskey Lake):
│     Device ID:          5792b48846ce271fab11c4a545f7a3df0d36e00a
│     Current version:    02
│     Vendor:             Intel Corporation (PCI:0x8086)
│     GUIDs:              b04f3e30-a317-5349-bbdd-c3c25153a5bd
│                         07616327-80e8-5a8a-9f8d-3cb40661abc4
│                         6986c21a-0f3f-5538-bcef-fe4e8cda7ab7
│                         52aa8ac7-c837-561f-af32-b7979cd2f6ed
│     Device Flags:       • Internal device
│                         • Cryptographic hash verification is available
│   
└─USB2.0 Hub:
      Device ID:          301046452a49d84af6356d23e43a684b8f10660f
      Summary:            USB 2.x Hub
      Current version:    51.13
      Vendor:             VIA Labs, Inc. (USB:0x17EF)
      Install Duration:   15 seconds
      GUIDs:              d5b585c2-e278-56ff-bd47-9bf0ad2bf9a9
                          8ee94f0e-9b44-596a-bdd9-6f90401664cc
                          0987e3c9-b1ee-5763-ac6e-51329b034e4b
                          0488c1ec-e9d6-589f-8620-f0a7426a3c4e
      Device Flags:       • Updatable
                          • Cryptographic hash verification is available
                          • Device stages updates
                          • Device can recover flash failures
```

**Additional questions**
- Operating system and version: Fedora 33
- Have you tried rebooting? Yup
- Is this a regression? No, it's my first time trying to update the docking station.

**Extra additional questions**
- what Lenovo power supplies you have plugged in: Lenovo 135W power adapter which came with the docking station: Model: ADL135NLC2A
- what USB devices and monitors were in the dock at the time of the upgrade: only a pendrive, 135W power adapter, and the usb-c connection from the docking station to the laptop. usb-c connection on the docking station is done at the back where it is supposed to be connected. No monitors connected.
- Laptop: Lenovo Thinkpad T490s, 32GB RAM, 256GB NVMe storage (standard RH company issue)

**Comment by hughsie** [comment]
@memily hey -- does this look like something you've seen before? I have the same hardware here but can't reproduce.

**Comment by SherlockChu219** [comment]
We reckon this might be a hardware issue. Please test these two methods for us:
1. Insert a pen drive to the hub and try updating with the pen drive attached.
2. Repeat fwupd update several times and record whether or not the error repeats in the same address 0x11be0 in the SPI Flash. 

Thanks.

**Comment by hughsie** [comment]
@thescouser89 would you be able to try the two things kindly requested from @SherlockChu219 please. Also, any more detail about the hardware would be useful, e.g. what Lenovo power supplies you have plugged in, what USB devices and monitors were in the dock at the time of the upgrade etc. Thanks!

**Comment by thescouser89** [comment]
Will do!

With pen drive inserted:
```
~ ❯ fwupdmgr update                                                      ✘ 1 9s
Devices with no available firmware updates: 
 • UEFI Device Firmware
 • UEFI Device Firmware
 • UEFI Device Firmware
 • UEFI dbx
 • USB2.0 Hub
Devices with the latest available firmware version:
 • Thunderbolt host controller
Upgrade available for USB3.1 Hub from 51.13 to 51.63
USB3.1 Hub and all connected devices may not be usable while updating. Continue with update? [Y|n]: Y
Downloading…             [***************************************]
Decompressing…           [***************************************]
Authenticating…          [***************************************]
Authenticating…          [***************************************]
Updating USB3.1 Hub…     [***************************************]
Erasing…                 [*****                                  ] Less than oneErasing…                 [********                               ] Less than oneErasing…                 [*****Erasing…                 [***************************************]…
failed to erase FW sector @0x15000: failed to read back empty: failed to read SPI data @0x159a0: transfer failed
```

I tried to put the pendrive both in front, and at the back if that matters.

The address sector is always different:
- failed to erase FW sector @0x10000: failed to read back empty: failed to read SPI data @0x104e0: transfer failed
- failed to erase FW sector @0x15000: failed to read back empty: failed to read SPI data @0x159a0: transfer failed
- failed to erase FW sector @0x13000: ->spi_write_status failed: failed to write SPI status 0x0: transfer failed
- failed to erase FW sector @0x12000: failed to read back empty: failed to read SPI data @0x12560: transfer failed
- failed to erase FW sector @0x11000: failed to read back empty: failed to read SPI data @0x11720: transfer failed

I'll update my initial description with the more detailed hardware info @hughsie .

Thanks guys for your help in figuring out the issue

**Comment by hughsie** [comment]
> transfer failed

Given all the address are pretty random, and we've even seen a write failure, I wonder if we need to retry some part of the control transfer on error? e.g.

	return g_usb_device_control_transfer (fu_usb_device_get_dev (FU_USB_DEVICE (self)),
					      G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
					      G_USB_DEVICE_REQUEST_TYPE_VENDOR,
					      G_USB_DEVICE_RECIPIENT_DEVICE,
					      0xc4, value, index,
					      buf, bufsz, NULL,
					      FU_VLI_DEVICE_TIMEOUT,
					      NULL, error);

Or maybe we need to do the fu_vli_device_get_spi_cmd() command and then do the control transfer read again?

...or perhaps the timeout is too small? 3000ms should be more than enough, right?

**Comment by SherlockChu219** [comment]
>I wonder if we need to retry some part of the control transfer on error? e.g.

We do not think the problem is software in nature. So we doubt retrying control transfer or adding a 3000ms timeout will address the issue because a read command only takes a few ms on U3. 

Given how the addresses are random, I assumed signal quality might be the problem...

If so then there are two types of quality fail:
1. USB signal connectivity issue related to the cable, receptor, or daisy chaining. 
2. SPI interface issue which on the PCB. 
 
We don't think it is possible to resolve this problem without contacting Lenovo customer service or having a Lenovo Repair center take a look at the device.

**Comment by hughsie** [comment]
> USB signal connectivity issue related to the cable

@thescouser89 are you using the original cable supplied with the Dock? If so, could you try a different USB cable? It doesn't have to be a thunderbolt cable for the USB hub update.

> SPI interface issue which on the PCB

Plausible. Would the Windows dock upgrade tools show anything like the same level of detail on failure? I wondered if the reporter should try the windows tool and see if the same error is generated.

**Comment by thescouser89** [comment]
@hughsie yup original cable.

I tried with another usb-c cable (most likely not thunderbolt) and I was able to update my audio firmware! I guess because it's not thunderbolt, I wasn't able to update the thunderbolt controller. I guess that's a cable issue then?

Unfortunately I won't get a hold of another usb-c thunderbolt cable until we're allowed back in the office. I wonder if Lenovo is willing to send me one for free... :)

Anyhow, I guess that answers why I was having issues. Thanks for the investigation!
